### DISCUSSION TITLE
Most of the training dataset words don't have a vector in Chinese

### DISCUSSION THREAD

**Discussion by ktobah** [explosion/spaCy#8886] [Help: Coding & Implementations] [answered]
I am trying to train a NER model with custom labels on a Chinese dataset (Spacy 3.1.1). I used label-studio for labeling and exported the dataset as .conll format, which I converted to Spacy binary format using the CLI convert command.

I tried to train and I could see that my loss is large and score very low. So, I tried to debug the data to see if there are any issues. I got the following:

I tried the default "char" tokenizer and "pkuseg" with the mixed model as well in my config file, but I got similar results.

```
root@guest:/media/newhd/ahmed/spacy_model# python3 -m spacy debug data config.cfg

============================ Data file validation ============================
✔ Corpus is loadable
✔ Pipeline can be initialized with data

=============================== Training stats ===============================
Language: zh
Training pipeline: tok2vec, ner
1 training docs
1 evaluation docs
✔ No overlap between training and evaluation data
✘ Low number of examples to train a new pipeline (1)

============================== Vocab & Vectors ==============================
ℹ 64236 total word(s) in the data (60504 unique)
ℹ 500000 vectors (500000 unique keys, 300 dimensions)
⚠ 62769 words in training data without vectors (98%)

========================== Named Entity Recognition ==========================
ℹ 122 label(s)
0 missing value(s) (tokens with '-' label)
⚠ Low number of examples for label '垃圾车' (37)

******

✔ No entities consisting of or starting/ending with whitespace
✔ No entities consisting of or starting/ending with punctuation

================================== Summary ==================================
✔ 5 checks passed
⚠ 109 warnings
✘ 1 error
```

98% of the words don't have a vector which I suspect is affecting the training.

Any idea, please, on how to solve this?

**Comment by polm** [comment]
Can you load up one of your training documents, iterate over the words, and see what percentage of them have no vectors? Something like this:

```
import spacy
nlp = spacy.load("zh_core_web_md")

text = ...

count = 0
doc = nlp(text)
for word in doc:
    if word.is_oov:
        count += 1
print(count / len(doc))
```

This should be the same as `debug data`, but you can print the document here to check that the tokenization is normal.

At first I thought that maybe the vector settings were just not obvious, but looking at your output this line in particular looks really weird:

    ℹ 64236 total word(s) in the data (60504 unique)

This suggests to me that something is wrong with your data, maybe an encoding issue or something. I would check that normal tokenization is working as expected outside training, as this seems worse than just a difference in tokenizer settings.

**Comment by adrianeboyd** [comment]
You might be seeing side effects related to #8774 if the problems were just with `spacy debug data`, but if there are also problems training, then it's probably something else in addition.

If you create a config just with the tokenizer (no additional components) and create a model from that config with `spacy assemble`, can you check the `TOK` value for your training data with `spacy evaluate model_dir train.spacy`?

**Comment by ktobah** [comment, marked as answer]
@polm and @adrianeboyd thank you for your suggestions.

Looking at a sample of my conll data file shows this:

```
-DOCSTART- -X- O
标的车直行碰撞,本车(***,*****).可以正常行驶.物损1(三者)其他物损(护栏)受损,不需要赔偿.无需报警,非现场. -X- _ B-护栏

标的车直行碰撞,本车(***,*****).可以正常行驶.物损1(三者)其他物损(石墩)受损,不需要赔偿.已建议报警,非现场. -X- _ B-石墩
```

When I googled, it seemed as if my text in the conll file is wrong. Mine includes the whole text while the ones I saw online include the tokens. Am I right?

@polm I would like to try what you suggested, but I am not able to find out how to read a .spacy file in the documentation. Can you please point me to where or how I can do it?

@adrianeboyd I would explore the data first to make sure it is not the issue here, then explore your suggestion. However, I do have the same issue with spacy train. I mean the scores are awful.

**Reply by polm** [reply]
> When I googled, it seemed as if my text in the conll file is wrong. Mine includes the whole text while the ones I saw online include the tokens. Am I right?

In a CoNLL file there should be one **token** per line. In some kinds of CoNLL files you will have a full sentence on a line, but only as a comment, the main data is always one token per line. Look at the [sample data](https://github.com/explosion/spaCy/tree/master/extra/example_data/ner_example_data). 

What's happening is each line is being treated as a token and so unsurprisingly whole sentences are not in the vectors table.

> I would like to try what you suggested, but I am not able to find out how to read a .spacy file in the documentation. Can you please point me to where or how I can do it?

You can use your original text in whatever format for the test I described, though I don't think you need to now that we know why your tokens are weird. If you want to use your `.spacy` file, that's just a serialized DocBin - see the [DocBin docs][docbin].

[docbin]: https://spacy.io/api/docbin#from_disk

**Reply by ktobah** [reply]
Thank you very much. It is clear now what the problem is. I will work to see why the labeling tool is saving it this way!
